### ISSUE TITLE
Yolo very slow in Android

### ISSUE THREAD

**Issue by Ngosti2000** [deeplearning4j/nd4j#2617]
Yolo2 outputSingle is very slow in Android to tunes of 8secs

My Application capturing Frames from the Android Camera and feeding them to TinyYoloModel... The response in terms of outputting the predictions is very slow around 8secs for a single prediction.

The expected behavior is to have the predictions below a second for the frames being fed into the YoloOutput Model

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j **0.9.2-SNAPSHOT**
* platform information (OS, etc) - **Android Mashmellow ,
Android 6.0.1 (Marshmallow),Chipset | Mediatek MT6755 Helio P10
CPU | Octa-core (4x2.0 GHz Cortex-A53 & 4x1.0 GHz Cortex-A53)**

Below is the Code 
 ```
public static List<DetectedObject> loadImage(Bitmap f) throws IOException {
        NativeImageLoader ni=new NativeImageLoader(416,416,3);      
        INDArray in= ni.asMatrix(f);
        if(yoloModel==null)return null;
        ImagePreProcessingScaler ips = new ImagePreProcessingScaler(0, 1);
        ips.transform(in);
        Yolo2OutputLayer outputLayer = (Yolo2OutputLayer) yoloModel.getOutputLayer(0);
        long start=System.currentTimeMillis();
        System.out.println("Start....."+start);
        INDArray output = yoloModel.outputSingle(in);
        long end =System.currentTimeMillis();
        System.out.println("End....."+end);
        System.out.println((end-start)/1000); //output is around 8 secs        
        List<DetectedObject> predictedObjects = outputLayer.getPredictedObjects(output, 0.6f);
        System.out.println((end-start)/1000);


```

The Profiler information is captured in the image below
![cuda](https://user-images.githubusercontent.com/290731/36031883-4a92512c-0dbd-11e8-8b99-43c59d8c39c3.png)

**Comment by raver119** [comment]
1) Please add screenshots that'll show us full picture. 370 microseconds spent on im2col isn't any news for us.
2) please check logger output on app startup. how many cores was reported? In your case it should be 2-4 ideally.

**Comment by Ngosti2000** [comment]
![screenshot from 2018-02-09 18-12-07](https://user-images.githubusercontent.com/290731/36034355-e86c4ea0-0dc4-11e8-8e65-ae859161494c.png)
![screenshot from 2018-02-09 18-15-12](https://user-images.githubusercontent.com/290731/36034464-373e4b1e-0dc5-11e8-8e3a-dbcbcfbe7a63.png)
![screenshot from 2018-02-09 18-17-40](https://user-images.githubusercontent.com/290731/36034573-90e9ffbe-0dc5-11e8-9b5c-4423c38a55a5.png)
![screenshot from 2018-02-09 18-20-34](https://user-images.githubusercontent.com/290731/36034709-f8affc48-0dc5-11e8-901d-27d772f7085d.png)
![screenshot from 2018-02-09 18-23-05](https://user-images.githubusercontent.com/290731/36034819-508b8054-0dc6-11e8-9f00-73239ed05895.png)
![screenshot from 2018-02-09 18-25-43](https://user-images.githubusercontent.com/290731/36035005-bd3b1944-0dc6-11e8-9693-ae64864090ee.png)
![screenshot from 2018-02-09 18-29-25](https://user-images.githubusercontent.com/290731/36035166-2f3995ca-0dc7-11e8-8469-831fffb35fbe.png)

**Comment by raver119** [comment]
Thank you man!

**Comment by Ngosti2000** [comment]
Below are the timings for the different configurations of the OMP_NUM_THREADS parameter
 /** OMP_NUM_THREADS -> Time in Secs
                 *  1-> 17/18/19/20
                 *  2-> 12/13
                 *  3-> 11/12
                 *  4-> 10/11
                 *  5-> 10/11
                 *  6-> 10/11
                 *  7-> 10/11
                 *  8-> 10/11
                 */

**Comment by raver119** [comment]
Thanks.

Could you please also post TF graph you've mentioned earlier? I'd like to compare implementations and data types used there

**Comment by Ngosti2000** [comment]
The TF graph used is SSD MobileNet. Here is the link to the paper https://arxiv.org/pdf/1611.10012.pdf and also for the TensorFlow source code on Android if it will be of any help https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android

**Comment by raver119** [comment]
just to confirm - you've been running TF example in YOLO mode, right?

**Comment by Ngosti2000** [comment]
Yes

**Comment by Ngosti2000** [comment]
@raver119 Any insight on the above issue?

**Comment by agibsonccc** [comment]
@Ngosti2000 raver was out sick today, he'll get back to you soonish. Sorry about that! It's been a rough week.

**Comment by Ngosti2000** [comment]
@agibsonccc Thanks for the update....... @raver119  hoping you a quick recovery

**Comment by Ngosti2000** [comment]
@raver119 ....any update?.... Am thinking maybe it would be better to adopt a  smaller model like MobileNet or SqueezeNet for Mobile as the weights are smaller in size Approx .17mb as compared to imagenet ~196MB and the accuracy is pretty much good. If you managed a hack for the imagenet pretrained model it will be great...

**Comment by raver119** [comment]
Working on that right now.

**Comment by saudet** [comment]
64-bit might be fix this? Try them out (the `android-arm64` and `android-x86_64` platforms)
